### ISSUE TITLE
Concurrency issue on graph startup

### ISSUE THREAD

**Issue by pietermartin** [pietermartin/sqlg#162]
Loading the topology in `Topology.cacheTopology` is not synchronized.
Starting up many graphs simultaneously causes severe errors as the `Topology` gets out of sync.

**Comment by pietermartin** [comment]
I just pushed a fix.
Can you check on your side that its behaving?

**Comment by JPMoresmau** [comment]
Ouch, no. TestMultiThread fails.
```
java.lang.AssertionError: expected:<200> but was:<139>
	at org.umlg.sqlg.test.schema.TestMultiThread.testMultipleGraphs(TestMultiThread.java:280)
...
java.lang.RuntimeException: org.postgresql.util.PSQLException: ERROR: deadlock detected
  Detail: Process 3099 waits for AccessExclusiveLock on relation 6482788 of database 18867; blocked by process 3183.
Process 3183 waits for AccessShareLock on relation 6482777 of database 18867; blocked by process 3099.
  Hint: See server log for query details.
	at org.umlg.sqlg.util.SqlgUtil.dropDb(SqlgUtil.java:816)
	at org.umlg.sqlg.test.BaseTest.before(BaseTest.java:94)

Caused by: org.postgresql.util.PSQLException: ERROR: deadlock detected
  Detail: Process 3099 waits for AccessExclusiveLock on relation 6482788 of database 18867; blocked by process 3183.
Process 3183 waits for AccessShareLock on relation 6482777 of database 18867; blocked by process 3099.
  Hint: See server log for query details.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2455)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2155)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:288)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:430)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:356)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:168)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:135)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:1423)
	at org.umlg.sqlg.util.SqlgUtil.dropDb(SqlgUtil.java:777)
	... 36 more


```

**Comment by JPMoresmau** [comment]
https://github.com/JPMoresmau/sqlg/tree/bugfix/recover_errors_multithread

Was an attempt to just recover on error. There is also a test that use a different label for each thread, to see if the topology loading was robust enough. The tests pass in that branch, but of course it's a bit hacky to just catch exception and retry.

**Comment by pietermartin** [comment]
Ha, have not had dead lock problems for a long time. They are bad!

**Comment by pietermartin** [comment]
I just ran the `TestMultiThread` tests on your branch and all 6 passed.
What //TODO?

**Comment by pietermartin** [comment]
ok sorry, my fault, tested master branch, getting failures now

**Comment by pietermartin** [comment]
Ok I am a tad confused. Your `bugfix/recover errors multithread` branch does not have the `Fix #163 ... lock` fixed merged into it.
BTW swallowing the errors won't work as the transaction is aborted serverside anyways.

I see snap-ci failed but I think its a timeout issue on the `executorService.awaitTermination` trying a longer timeout.

**Comment by pietermartin** [comment]
@JPMoresmau can you test master branch again on your system?
I included the multi label test you added.

snap-ci was failing but no exception just assertion failures. I suspect postgresql's max_connection is the bottle neck. Anyhow snap-ci is discontinued so I do not want to spend time on its settings.
Tried travis again, it fails with disc out of space errors. Travis is disappointing.
Tried codeship and it is passing with default settings.

My local postgres is on 1000 max_connections in postgresql.conf.
However I tried with 100 and the tests still passed.

If your `TestMultiThread.testMultipleGraphs`  is failing perhaps try increase the max_connections setting and see if it makes a difference.

Thanks

**Comment by JPMoresmau** [comment]
@pietermartin 
1. Yes, my branch was done in parallel to your own work so didn't include your fixes. I think we can ignore that branch now that you've implemented the multi label test
2. I was using 9.5, I've now upgraded to 9.6 (which is what we're using in work) and set max_connections to 1000 -> TestMultiThread passes 100%.

So we're good to go with master! Thanks!

**Comment by pietermartin** [comment]
Great, tx
